### PR TITLE
ddl: make DDLJob.Query using "show create table" sql when create clustered table with session vars

### LINKED ISSUE #24796: Binlog/CDC build nonclustered table when create clustered table with session var in upstream
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. setup two cluster a and b, and replica a to b by binlog or cdc
2. set tidb_enable_clustered_index=1, create a clustered table without clustered keyword in upstream cluster a
3. show create table in cluster b

### 2. What did you expect to see? (Required)

new table in cluster b should be CLUSTERD

### 3. What did you see instead (Required)

NONCLUSTERD

### 4. What is your TiDB version? (Required)

master, 5.0.1

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: close #24796 

Problem Summary:

first, simple introduce how tidb/cdc/binlog works with create table with clustered

1. TiDB setup a tableInfo with `IsCommonHandle = true`
2. TiDB enqueue DDLJob with origin query --- `create table t (a char(1) primary key)` 
3. For Binlog,  tidb will replace ddlJob.Query's keyword as hint in https://github.com/pingcap/tidb/blob/18cbfaac15f8478902726e1e64971aa96b862462/sessionctx/binloginfo/binloginfo.go#L285, then output hinted-query in binlogInfo
4. For CDC,  cdc will subscribe tidb's ddlJob and replace ddlJob.Query to hinted query in  https://github.com/pingcap/ticdc/blob/7346b76d427dfa6717eb2b56e3872a7c1e53bdf8/cdc/changefeed.go#L706.

so, for `set tidb_enable_clustered_index=1` + `create table t (a char(1) primary key)` can not be hinted in step3 or step4, because it's not keyword, only a session vars.

### What is changed and how it works?

What's Changed, How it Works:

choose use `show create table` result in step2's ddlJob.Query

`StatementContext.UseShowCreateInDDLJob` and `CtorShowCreateTableHook` is very dirty, but it seems the cheapest way - 

> maybe we also can choose rewrite query like `AddSpecialComment` but use regexp to replace a SQL is very dangerous(e.g. `create table t5(`AUTO_RANDOM_BASE=3` int auto_increment key)` should cause downstream use different column name)

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`: will do it with clustered binlog support
- Need to cherry-pick to the release branch 5.1

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

```
setup two clusters with  cdc or binlog, create a clustered index table in upstream and check it in downstream
```

Side effects

- n/a

### Release note 

- Will do it with clustered binlog support release item

### CODE REVIEW

**Comment by lysu** [comment]
@tangenta @wjhuang2016 @wangggong  PTAL thx

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1884,13 +1889,23 @@ func (d *ddl) CreateTableWithInfo(
...
 	}
 
+	createTableQuery, _ := ctx.Value(sessionctx.QueryString).(string)
+	if createTableQuery != "" && ctx.GetSessionVars().StmtCtx.UseShowCreateInDDLJob {
+		var buf bytes.Buffer
+		err = CtorShowCreateTableHook(ctx, tbInfo, nil, &buf)
+		if err != nil {
+			return err
+		}
+		createTableQuery = buf.String()
```
Comment (AilinKid) [inline ddl/ddl_api.go:1899]: CreateTableWithInfo in the same session will continuously walk through this, should set `UseShowCreateInDDLJob` back?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -1821,6 +1823,9 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 	return d.CreateTableWithInfo(ctx, schema.Name, tbInfo, onExist, false /*tryRetainID*/)
 }
 
+// CtorShowCreateTableHook is the hook to call `executor.ConstructResultOfShowCreateTable` from ddl package(or it will cycle dependence)
```
Comment (AilinKid) [inline ddl/ddl_api.go:1826]: s/CtorShowCreateTableHook/CtlShowCreateTableHook

File: executor/show.go [hunk 3]
```diff
@@ -969,32 +973,36 @@ func ConstructResultOfShowCreateTable(ctx sessionctx.Context, tableInfo *model.T
 		fmt.Fprintf(buf, " COMPRESSION='%s'", tableInfo.Compression)
 	}
 
-	incrementAllocator := allocators.Get(autoid.RowIDAllocType)
-	if hasAutoIncID && incrementAllocator != nil {
-		autoIncID, err := incrementAllocator.NextGlobalAutoID(tableInfo.ID)
-		if err != nil {
-			return errors.Trace(err)
-		}
+	if allocators != nil {
```
Comment (wjhuang2016) [inline executor/show.go:976]: Why do you update this?

File: ddl/ddl_api.go (hunk 1 above)
Comment (wjhuang2016) [reply ddl/ddl_api.go:1899]: Can we always use the show create table?

File: executor/show.go (hunk 3 above)
Comment (tangenta) [reply executor/show.go:976]: Because the code in upper level passes in a nil `allocators`...

**Comment by tangenta** [comment]
This is sooo dirty. How can the other session variables that affect DDL be handled in this case? For example, `@@session.tidb_enable_list_partition`, `@@session.tidb_enable_change_column_type`...

**Comment by wjhuang2016** [comment]
> This is sooo dirty. How can the other session variables that affect DDL be handled in this case? For example, `@@session.tidb_enable_list_partition`, `@@session.tidb_enable_change_column_type`...

I think they are not the same. Since `@@session.tidb_enable_list_partition`, `@@session.tidb_enable_change_column_type` doesn't affect downstream replication

**Comment by wangggong** [comment]
LGTM

**Pull request closed without merging**